src/cmap/conn/mod.rs

```diff
@@ -207,9 +211,11 @@ impl Connection {
         request_id: impl Into<Option<i32>>,
     ) -> Result<CommandResponse> {
         let message = Message::with_command(command, request_id.into());
-        message.write_to(&mut self.stream).await?;
+        let request_id = message.write_to(&mut self.stream).await?;
+        self.pending_request_ids.push_back(request_id);
 
         let response_message = Message::read_from(&mut self.stream).await?;
+        self.pending_request_ids.pop_front();
```

Comment:
My thought was that you would discard irrelevant results here, rather than in your drop implementation. I don't think you need to keep around the `pending_request_ids` in that case, rather just `read_from` until you get the id you're looking for or timeout. I think this "lazy" approach is easier to reason about, tracks less state, and has the added benefit of not doing I/O in your drop (maybe think of it as "only pay for the I/O you need").

Reply:
The main reason I'm not super thrilled with doing it here is that it offloads the cost of reading the responses to the next operation. Ideally, if a user drops the future for an operation before reading the response, the next operation wouldn't have to pay any performance penalty for that. I think it's worth making the distinction this implementation doing do I/O in the drop implementation; it does it in the background task that the drop operation happened to have started. The actual code executed when the connection is dropped is still the same; the runtime will just schedule some extra work to be done at some point later without holding up the code path where the drop occurs.

That being said, cleaning it up before sending a new message does guarantee that there won't ever be more than one message stuck in the buffer waiting to be read, so the cost of reading it before sending a new message likely won't be that expensive compared to the network round trip that has to be done to actually perform the operation. I think there's a little bit of trickiness with regards to if the next operation's future happens to also be dropped, but before the buffer is fully flushed, so I'll add a bit more logic to handle that case as well.